File: apps/frontend/src/api/market.ts
import { z } from "zod";
import { api, type ApiEnvelope } from "../lib/api";

export const SymbolsSchema = z.object({
  symbols: z.array(z.string()),
});
export type SymbolsResponse = z.infer<typeof SymbolsSchema>;

export const BarSchema = z.object({
  symbol: z.string(),
  interval: z.string(),
  open_time: z.number().int(),
  close_time: z.number().int(),
  open: z.string(),
  high: z.string(),
  low: z.string(),
  close: z.string(),
  volume: z.string(),
});
export type Bar = z.infer<typeof BarSchema>;

export const CandlesSchema = z.object({
  symbol: z.string(),
  interval: z.string(),
  bars: z.array(BarSchema),
});
export type CandlesResponse = z.infer<typeof CandlesSchema>;

export const SnapshotSchema = z.object({
  symbol: z.string(),
  interval: z.string(),
  latest: BarSchema,
  source: z.string(),
});
export type SnapshotResponse = z.infer<typeof SnapshotSchema>;

async function unwrap<T>(
  path: string,
  schema: z.ZodSchema<T>,
  params?: Record<string, string | number>,
): Promise<T> {
  const { data } = await api.get<ApiEnvelope<unknown>>(path, { params });
  if (!data || data.error || data.data == null) {
    throw new Error(data?.error?.message ?? "no data");
  }
  return schema.parse(data.data);
}

export const getSymbols = (): Promise<SymbolsResponse> =>
  unwrap("/api/market/symbols", SymbolsSchema);

export const getSnapshot = (
  symbol: string,
  interval = "1m",
): Promise<SnapshotResponse> =>
  unwrap("/api/market/snapshot", SnapshotSchema, { symbol, interval });

export const getCandles = (
  symbol: string,
  interval = "1m",
  limit = 30,
): Promise<CandlesResponse> =>
  unwrap("/api/market/candles", CandlesSchema, { symbol, interval, limit });

File: apps/backend/internal/api/handlers/market.go
package handlers

import (
	"errors"
	"strconv"
	"strings"

	"github.com/gofiber/fiber/v2"

	"github.com/jakkrit-ch/trading-ai-v2/backend/internal/data"
	"github.com/jakkrit-ch/trading-ai-v2/backend/internal/data/market"
	"github.com/jakkrit-ch/trading-ai-v2/backend/internal/domain"
)

const (
	marketDefaultInterval = "1m"
	marketDefaultLimit    = 100
	marketMaxLimit        = 1000
)

type marketH struct {
	svc *market.Service
}

func newMarket(svc *market.Service) *marketH { return &marketH{svc: svc} }

func (h *marketH) symbols(c *fiber.Ctx) error {
	syms := h.svc.Symbols()
	out := make([]string, 0, len(syms))
	for _, s := range syms {
		out = append(out, string(s))
	}
	return writeOK(c, map[string]interface{}{"symbols": out})
}

// barDTO mirrors domain.Bar with decimals as strings.
type barDTO struct {
	Symbol    string `json:"symbol"`
	Interval  string `json:"interval"`
	OpenTime  int64  `json:"open_time"`
	CloseTime int64  `json:"close_time"`
	Open      string `json:"open"`
	High      string `json:"high"`
	Low       string `json:"low"`
	Close     string `json:"close"`
	Volume    string `json:"volume"`
}

func toBarDTO(b domain.Bar) barDTO {
	return barDTO{
		Symbol:    string(b.Symbol),
		Interval:  b.Interval,
		OpenTime:  b.OpenTime,
		CloseTime: b.CloseTime,
		Open:      b.Open.String(),
		High:      b.High.String(),
		Low:       b.Low.String(),
		Close:     b.Close.String(),
		Volume:    b.Volume.String(),
	}
}

func (h *marketH) candles(c *fiber.Ctx) error {
	sym, interval, ok := readMarketParams(c)
	if !ok {
		return nil
	}
	limit := marketDefaultLimit
	if v := strings.TrimSpace(c.Query("limit")); v != "" {
		n, err := strconv.Atoi(v)
		if err != nil || n <= 0 {
			return writeError(c, fiber.StatusBadRequest, "validation_failed", "limit must be a positive integer")
		}
		if n > marketMaxLimit {
			n = marketMaxLimit
		}
		limit = n
	}

	bars, err := h.svc.Candles(c.UserContext(), sym, interval, limit)
	if err != nil {
		return writeError(c, fiber.StatusInternalServerError, "internal", err.Error())
	}
	out := make([]barDTO, 0, len(bars))
	for _, b := range bars {
		out = append(out, toBarDTO(b))
	}
	return writeOK(c, map[string]interface{}{
		"symbol":   string(sym),
		"interval": interval,
		"bars":     out,
	})
}

func (h *marketH) snapshot(c *fiber.Ctx) error {
	sym, interval, ok := readMarketParams(c)
	if !ok {
		return nil
	}
	bar, source, err := h.svc.Snapshot(c.UserContext(), sym, interval)
	if err != nil {
		if errors.Is(err, data.ErrNotFound) {
			return writeError(c, fiber.StatusNotFound, "not_found", "no snapshot for symbol/interval")
		}
		return writeError(c, fiber.StatusInternalServerError, "internal", err.Error())
	}
	return writeOK(c, map[string]interface{}{
		"symbol":   string(sym),
		"interval": interval,
		"latest":   toBarDTO(bar),
		"source":   source,
	})
}

func readMarketParams(c *fiber.Ctx) (domain.Symbol, string, bool) {
	sym := strings.ToUpper(strings.TrimSpace(c.Query("symbol")))
	if sym == "" {
		_ = writeError(c, fiber.StatusBadRequest, "validation_failed", "symbol is required")
		return "", "", false
	}
	interval := strings.TrimSpace(c.Query("interval"))
	if interval == "" {
		interval = marketDefaultInterval
	}
	return domain.Symbol(sym), interval, true
}
